ipa-core/src/protocol/ipa_prf/malicious_security/lagrange.rs

```diff
@@ -0,0 +1,240 @@
+use std::iter;
+
+use generic_array::{ArrayLength, GenericArray};
+use typenum::U1;
+
+use crate::ff::{Field, PrimeField};
+
+/// A degree `N-1` polynomial is stored as `N` points `(x,y)`
+/// where the "x coordinates" of the input points are `x_0` to `x_N` are `F::ZERO` to `(N-1)*F::ONE`
+/// Therefore, we only need to store the `y` coordinates.
+#[derive(Debug, PartialEq, Clone)]
+pub struct Polynomial<F: Field, N: ArrayLength> {
+    y_coordinates: GenericArray<F, N>,
+}
+
+/// The Canonical Lagrange denominator is defined as the denominator of the Lagrange base polynomials
+/// `https://en.wikipedia.org/wiki/Lagrange_polynomial`
+/// where the "x coordinates" of the input points are `x_0` to `x_N` are `F::ZERO` to `(N-1)*F::ONE`
+/// the degree of the polynomials is `N-1`
+pub struct CanonicalLagrangeDenominator<F: Field, N: ArrayLength> {
```

Comment:
we first need to compute the denominators, then we can compute the table. It makes sense to separate the two since the denominator only needs the input points while Lagrange table also needs the output points. So we can use the same denominator for several Lagrange tables. In the proofs, we need a table for the output points N,...2*N-1 as well as a table for each random challenge r (which is generated as the hash of the "subproof" of the previous iteration). The size of the denominator is likely to be around 8-32 field elements.